Chat with PDF using Chainlit

neelasha23 opened the pull request:
Closes #167 

<!-- readthedocs-preview ploomber-doc start -->
----
📚 Documentation preview 📚: https://ploomber-doc--178.org.readthedocs.build/en/178/

<!-- readthedocs-preview ploomber-doc end -->

edublancas commented:
does langchain provide any significant benefits? you could also do it with the openAI package but if you feel like using langchain is better, go ahead

neelasha23 commented:
> does langchain provide any significant benefits? you could also do it with the openAI package but if you feel like using langchain is better, go ahead

The embeddings are from OpenAI. langchain has features for making the conversation a contextual one. So the chat memory will append the current user query to the previous response and if there's any additional info in previous msg it takes that into account.  I referred to this tutorial: https://docs.chainlit.io/examples/qa

Should I try without langchain completely ?

@edublancas

edublancas commented:
that's ok, let's keep langchain

neelasha23 commented on examples/chainlit/chat-with-pdf/README.md:
```diff
@@ -0,0 +1,29 @@
+* Follow the instructions for deploying a [Chainlit](https://docs.cloud.ploomber.io/en/latest/apps/chainlit.html) application.
+* You also need to set a couple of [environment variables](https://docs.cloud.ploomber.io/en/latest/user-guide/secrets.html) while deploying the application.
+    * `OPENAI_API_KEY`
+    * `PLOOMBER_CLOUD_KEY`: Refer to [documentation](https://docs.cloud.ploomber.io/en/latest/quickstart/apikey.html)
```
I had to add this step because of this error:

<img width="1240" alt="Screenshot 2024-04-06 at 7 05 52 PM" src="https://github.com/ploomber/doc/assets/29096135/623b101e-12b1-4d68-ae29-677a573688dc">

bryannho reviewed (commented):
Code looks good. Will wait for deploy to test functionality

bryannho commented:
@neelasha23 I got this error when I uploaded a PDF (didn't send any messages yet):

![image](https://github.com/ploomber/doc/assets/42661667/515542aa-f7a3-42b2-9fbe-78f11d313433)

neelasha23 commented:
> @neelasha23 I got this error when I uploaded a PDF (didn't send any messages yet):

Please share the PDF @bryannho

bryannho commented:
@edublancas It was failing because I used a scanned PDF. 

@neelasha23 will add a catch and use [scannned_pdf_to_text](https://github.com/ploomber/serverless/blob/main/api/api.py#L176-L193) for this case

bryannho commented:
@neelasha23 two more small notes:

1. I tried again with this native [placeholder.pdf](https://github.com/ploomber/doc/files/14910537/placeholder.pdf) and it returned the same error `list index out of range`, so unsure if it's just a scanned pdf issue.
2. The "Processing placeholder.pdf..." message isn't sent until the document has fully finished processing which is a bit confusing. I see the progress bar for the file upload, but could you try to send that message **after** the file upload has completed but **before** the app starts parsing the document? See the video below:


https://github.com/ploomber/doc/assets/42661667/26042d13-d33e-4c32-8874-d26efd1d4299

neelasha23 commented:
> @edublancas It was failing because I used a scanned PDF.
> 
> @neelasha23 will add a catch and use [pdf_scanned_to_text]

Added call to [pdf_scanned_to_text](https://github.com/ploomber/doc/pull/178/commits/5d94150b8011c3014e180758262090ad00242d53)

@bryannho

neelasha23 commented:
1. I tried again with this native [placeholder.pdf](https://github.com/ploomber/doc/files/14910537/placeholder.pdf) and it returned the same error `list index out of range`

It happened because there was just one line in the PDF. [Fixed](https://github.com/ploomber/doc/pull/178/commits/c3c75520350ff3e84152b105e8ff7004b13624c9#diff-510fae8e6971b8c3fb6d20125723a926993e6f9a765df63a133202f348bd42a1R80)

> 2. could you try to send that message **after** the file upload has completed but **before** the app starts parsing the document? 

It was already after the file upload and before doc parsing. But I think because of the asynchronous nature of calling pdf_to_text the message was getting stuck. Have convrted the pdf conversion call to [async](https://github.com/ploomber/doc/pull/178/commits/c3c75520350ff3e84152b105e8ff7004b13624c9#diff-510fae8e6971b8c3fb6d20125723a926993e6f9a765df63a133202f348bd42a1R26-R36)

@bryannho @edublancas

edublancas commented:
the answers look off (I tried this paper:  https://arxiv.org/abs/2402.00838):

![image](https://github.com/ploomber/doc/assets/989250/e1239c1d-d3c2-4477-85ce-94de739c4e82)

neelasha23 commented:
> the answers look off (I tried this paper: https://arxiv.org/abs/2402.00838):
> 

I think it's confusing the title with the ones in the References section. 

<img width="564" alt="Screenshot 2024-04-09 at 10 11 37 PM" src="https://github.com/ploomber/doc/assets/29096135/7234f68c-398e-408b-bbe7-c17662c8744c">

It is answering specific questions about the content though:

<img width="965" alt="Screenshot 2024-04-09 at 10 13 01 PM" src="https://github.com/ploomber/doc/assets/29096135/ed23f52b-c115-420d-aa60-65b59b4d618c">

@edublancas

edublancas commented:
I think we should fix it, those basic questions should be answered correctly.

check how it's done here: https://github.com/ploomber/doc/tree/main/examples/panel/chat-with-pdf I remember testing that example and it worked fine when asking about the abstract and title

two things might happen:

- the issue might be in how we pass the info to langchain, in such case, check if you want to change the settings
- maybe the problem is the content returned by our parsing functions. in such case, let us know so we can see how we can improve it

neelasha23 commented:
The Panel one is also mistaking the title:

<img width="908" alt="Screenshot 2024-04-09 at 10 24 10 PM" src="https://github.com/ploomber/doc/assets/29096135/1281abc1-a862-4e19-82de-8416258d001f">

But the 2nd response is better:

<img width="959" alt="Screenshot 2024-04-09 at 10 26 28 PM" src="https://github.com/ploomber/doc/assets/29096135/51d13de5-5c79-4af2-85dd-212628eda4e0">

I'll still check what can be improved

@edublancas

edublancas commented:
cool. yeah, compare the answers from both examples, and try to see if you can improve it a bit.

spend ~2 hours on this; no need to spend more. let me know your conclusions so we know if we should keep digging or publish as is

neelasha23 commented:
I ended up spending a bit more time on this as even after adding the same settings as the Panel app the results were not comparing with the outputs I saw in the Panel app yesterday. But I tested the Panel app again and I realized it's not performing any better. The results are inconsistent. It might give the correct answer once in a while by chance. Also, this issue is mostly with the OLMO paper. I found better results with other papers. Below are the detailed observations and next steps:

**Changes made to Chainlit App** (as per the Panel one)

- Changed `RecursiveCharacterTextSplitter` to `CharacterTextSplitter`
- Converted the PDF extracted using Ploomber API to a list of [Document](https://python.langchain.com/docs/modules/data_connection/document_loaders/) objects. Each object looks like this:
  `Document(page_content="text", metadata={"source": page_path, "page": page number})`
- As we now have `Document` objects converted `LanceDB.from_texts` to `LanceDB_from_documents`.
- Converted `ConversationalRetrievalChain` to `RetrievalQA`
- Added `search_type="similarity"` 

**Panel app observations** (inconsistent results on the Olmo paper)
<img width="978" alt="Screenshot 2024-04-10 at 4 53 09 PM" src="https://github.com/ploomber/doc/assets/29096135/81f1abd9-2665-460b-984f-584a30cec6aa">

It has picked up from the references:
<img width="815" alt="Screenshot 2024-04-10 at 4 54 14 PM" src="https://github.com/ploomber/doc/assets/29096135/4dd869b6-dc35-475b-b8f0-3c43448b26a1">

Another attempt:

<img width="1031" alt="Screenshot 2024-04-10 at 4 59 24 PM" src="https://github.com/ploomber/doc/assets/29096135/681d740d-0c24-4242-82bc-4a2f5d24f926">

**Chainlit output**

Paper: https://proceedings.neurips.cc/paper_files/paper/2017/file/3f5ee243547dee91fbd053c1c4a845aa-Paper.pdf

<img width="1051" alt="Screenshot 2024-04-10 at 5 20 35 PM" src="https://github.com/ploomber/doc/assets/29096135/b2778aa8-d298-45f4-9fda-5a70fe5af6b7">


**Conclusions**

1. It looks like in both cases the vector search is biased towards the `References` section, which might be because it's towards the end of the document.
2. We can research more on how to utilize the `metadata` in the `Document` object to improve the results.
3. We can explore a RAG-based approach and index different parts of the document separately so the retriever doesn't get confused.
4. The `pdf_to_text` and `scanned_pdf_to_text` functions can have a parameter for returning results in the `Document` format mentioned above. That would make it compatible with `langchain` (in case it's relevant for users)
5. I think the vector DB should not make too much of a difference but we can check differences between `LanceDB` and `Chroma` (that's the only difference now in the Chainlit app vs Panel app). Although I did try with Chroma also but it didn't make any difference.

Deployed updated app: https://www.platform.ploomber.io/applications/steep-salad-8357/0f7566de

@edublancas

edublancas commented:
@neelasha23 please prepare social media posts and include the link to the app, let's keep the example running for a week
